components/sup/src/events.rs

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Main interface for a stream of events the Supervisor can send out
+//! in the course of its operations.
+//!
+//! Currently, the Supervisor is able to send events to a [NATS
+//! Streaming][1] server. The `init_stream` function must be called
+//! before sending events to initialize the publishing thread in the
+//! background. Thereafter, you can pass "event" structs to the
+//! `event` function, which will publish the event to the stream.
+//!
+//! All events are published under the "habitat" subject.
+//!
+//! [1]:https://github.com/nats-io/nats-streaming-server
+
+mod nats;
+mod types;
+
+use futures::sync::mpsc::UnboundedSender;
+pub use nats::EventConnectionInfo;
+use state::Container;
+use std::{fmt::Debug,
+          sync::Once};
+pub use types::*;
+
+static INIT: Once = Once::new();
+lazy_static! {
+    // TODO (CM): When const fn support lands in stable, we can ditch
+    // this lazy_static call.
+
+    /// Reference to the event stream.
+    static ref EVENT_STREAM: Container = Container::new();
+    /// Core information that is shared between all events.
+    static ref EVENT_CORE: Container = Container::new();
+}
+
+/// Starts a new thread for sending events to a NATS Streaming
+/// server. Stashes the handle to the stream, as well as the core
+/// event information that will be a part of all events, in a global
+/// static reference for access later.
+pub fn init_stream(conn_info: EventConnectionInfo, event_core: EventCore) {
+    INIT.call_once(|| {
+        let event_stream = nats::init_stream(conn_info).expect("Could not start NATS thread");
+        EVENT_STREAM.set(event_stream);
+        EVENT_CORE.set(event_core);
+    });
+}
+
+/// Publish an event. This is the main interface that client code will
+/// use.
+///
+/// If `init_stream` has not been called already, this function will
+/// be a no-op.
+// NOTE: we can take advantage of this to "disable" the event
+// subsystem if users don't wish to send events out; just don't call
+// `init_stream` if they don't want it.
+pub fn event<E>(event: E)
+where
+    E: Event,
+{
+    // TODO: incorporate the current timestamp into the rendered event
+    // (which will require tweaks to the rendering logic, but we know
+    // that'll need to be updated anyway).
+
+    // We render the event to bytes here, rather than over in the
+    // publication thread, because it allows our Event types to deal
+    // with references, which means less allocations and unnecessary
+    // copying (otherwise, we'd have to have copies of everything to
+    // send it over to the other thread, or much more complicated
+    // thread-safe types!). It also makes the publication thread logic
+    // a bit simpler; it just takes bytes and sends them out.
+    if let Some(e) = EVENT_STREAM.try_get::<EventStream>() {
+        e.send(event.render(EVENT_CORE.get::<EventCore>()));
+    }
```

Comment:
That would get called all the time when people don't have the event feature enabled.

I'd rather have the logging take place once at the point where we make the choice to enable or disable event streaming (once this is a real feature, and not hidden behind a feature flag, that is.)

Reply:
True, but it would only do anything when the logging level was set high enough that someone was trying to debug. Since this is so cache-friendly, I have a hard time imagining it's a performance concern. That said, if you think adding the *code* here is just too ugly and that the idea someone would be confused is too remote, I don't see it as a must-have.

Reply:
For now, I'd rather have the logging take place when we determine whether or not to enable the event stream. We can revisit if that ends up not being the right choice.